qlib/model/trainer.py

```diff
@@ -45,33 +45,35 @@ def begin_task_train(task_config: dict, experiment_name: str, recorder_name: str
     return recorder
 
 
-def fill_placeholder(kwargs, model, dataset):
+def fill_placeholder(config: dict, config_extend: dict):
     """
-        Detect placeholder(<MODEL> and <DATASET>) in dict and fill them.
+    Detect placeholder in config and fill them with config_extend.
 
-        Args:
-            kwargs (Dict): the parameter dict will be filled
-            model (Model): fill <MODEL>
-            dataset (Dataset): fill <DATASET>
+    Parameters
+    ----------
+    config : dict
+        the parameter dict will be filled
+    config_extend : dict
+        the value of all placeholders
 
-        Returns:
-            Dict: the parameter dict
-    """
+    Returns
+    -------
+    dict
+        the parameter dict
+    """    
     top = 0
     tail = 1
-    dict_quene = [kwargs]
+    dict_quene = [config]
     while(top < tail):
         now_dict = dict_quene[top]
         top += 1
         for key in now_dict.keys():
             if(isinstance(now_dict[key], dict)):
                 dict_quene.append(now_dict[key])
                 tail += 1
-            elif(now_dict[key] == "<MODEL>"):
-                now_dict[key] = model
-            elif(now_dict[key] == "<DATASET>"):
-                now_dict[key] = dataset 
-    return kwargs
+            elif(now_dict[key] in config_extend.keys()):
+                now_dict[key] = config_extend[now_dict[key]]
+    return config
```

Comment:
This function should be implemented in a recursive way for better generality.

Reply:
Here is a example with similar logic 
https://github.com/microsoft/qlib/blob/main/qlib/utils/__init__.py#L750

Reply:
why dfs is more general than bfs?